service/raft/raft_group_registry.cc

```diff
@@ -408,6 +413,98 @@ shared_ptr<raft::failure_detector> raft_group_registry::failure_detector() {
 
 raft_group_registry::~raft_group_registry() = default;
 
+namespace {
+    class composite_abort_source {
+        abort_source _as;
+        utils::small_vector<abort_source::subscription, 2> _subscriptions;
+    public:
+        void add(abort_source& as) {
+            as.check();
+            auto sub = as.subscribe([this]() noexcept { _as.request_abort(); });
+            assert(sub);
+            _subscriptions.push_back(std::move(*sub));
+        }
+        abort_source& abort_source() noexcept {
+            return _as;
+        }
+    };
+    auto fmt_loc(const seastar::compat::source_location& l) {
+        return fmt::format("{}({}:{}) `{}`", l.file_name(), l.line(), l.column(), l.function_name());
+    };
+}
+
+raft_server_with_timeouts::raft_server_with_timeouts(raft::server& server, raft_group_registry& registry,
+        std::optional<lowres_clock::duration> timeout)
+    : _server(server)
+    , _registry(registry)
+    , _timeout(timeout)
+{
+}
+
+template <std::invocable<abort_source*> Op>
+std::invoke_result_t<Op, abort_source*>
+raft_server_with_timeouts::run_with_timeout(Op&& op, const char* op_name,
+    seastar::abort_source* as, std::optional<raft_timeout> timeout)
+{
+    if (!timeout) {
+        co_await op(as);
+        co_return;
+    }
+    if (!timeout->value) {
+        if (!_timeout) {
+            on_internal_error(rslog, ::format("raft operation [{}], timeout requested at [{}],"
+                                              "but no value for it has been defined",
+                op_name, fmt_loc(timeout->loc)));
+        }
+        timeout->value = _timeout.value();
+    }
+    composite_abort_source composite_as;
+
+    abort_on_expiry<> expiry{lowres_clock::now() + *timeout->value};
+    composite_as.add(expiry.abort_source());
+
+    if (as) {
+        composite_as.add(*as);
+    }
+
+    try {
+        co_return co_await op(&composite_as.abort_source());
+    } catch (const raft::request_aborted& e) {
+        if (!expiry.abort_source().abort_requested() || (as && as->abort_requested())) {
+            throw;
+        }
+        // TODO: improve error message (check for quorum loss etc)
+        const auto message = ::format("raft operation [{}] timed out", op_name);
+        static logger::rate_limit rate_limit{std::chrono::seconds(1)};
+        rslog.log(log_level::warn, rate_limit, "{}; timeout requested at [{}], original error {}",
+            message, fmt_loc(timeout->loc), std::current_exception());
+        throw std::runtime_error(message.c_str());
```

Comment:
Right.

Probably it would be best to use something that's understandable to `cqlsh`. Because, if I remember correctly, `cqlsh` does not print the contents of `runtime_error`, it translates it into vague "server error". But here it would be nice to show that we don't have quorum, for example.

Could we use one of the error types that driver understands?